README.md
<img src="https://raw.githubusercontent.com/lacarmen/cryogen/master/cryogen.png"
 hspace="20" align="left" height="200"/>

<!-- Non-breaking space -->
&nbsp;

This README only documents a subset of Cryogen's features. For additional documentation please see the [cryogen site](http://cryogenweb.org).

<!-- Non-breaking space -->
&nbsp;


## Features

* Blog posts and pages with Markdown (default) or AsciiDoc
* Tags
* Table of contents generation
* Plain HTML page templates
* Code syntax highlighting
* Disqus support
* Sitemap generation
* RSS feed generation
* Sass/SCSS compilation
* Klipse Integration

## Usage

### Creating a New Site

You can create a new website using either one of `leiningen` or `clj-new` or `deps-new`.

#### Creating a New Site With Leiningen

You will need [Leiningen][1] 2.5.0 or above installed.

[1]: https://github.com/technomancy/leiningen

A new site can be created using the Cryogen leiningen template as follows:

```sh
lein new cryogen my-blog
```

#### Creating a New Site With clj-new as a Tool

Alternatively, use [`clj-new as a tool`](https://github.com/seancorfield/clj-new#installation-as-a-tool):

```sh
clojure -Ttools install com.github.seancorfield/clj-new '{:git/tag "v1.2.362"}' :as clj-new # update to latest!
clojure -Tclj-new create :template cryogen :name myname/myblog :force true
cd myname/myblog/
```

#### Creating a New Site With deps-new as a Tool

Alternatively, use [`deps-new as a tool`](https://github.com/seancorfield/deps-new#deps-new-):

```sh
clojure -Ttools install io.github.seancorfield/deps-new '{:git/tag "v0.4.0"}' :as new
clojure -Sdeps '{:deps {io.github.cryogen-project/cryogen {:git/tag "0.6.6" :git/sha "fcb2833"}}}' -Tnew :template org.cryogenweb/new :name myname/myblog
cd myname/myblog/
```

The artifact `cryogen/lein-template` contains both a leiningen
template and a deps-new template.

### Running the Server

The web server can be started from the `my-blog` directory using either Leiningen:

```sh
lein serve # or lein serve:fast
```

or tools-deps:

```sh
clojure -X:serve # or clojure -X:serve:fast
```

The server will watch for changes in the `content` and `themes` folders and recompile the content automatically. The `*:fast` variants perform [fast but partial compilation](https://cryogenweb.org/docs/fast-compilation.html) of only the changed page/post.

You can also generate the content without bringing up a server either via:

```sh
lein run
```

or via:

```sh
clojure -M:build
```

### Site Configuration

The site configuration file is found at `content/config.edn`, this file looks as follows:

```clojure
{:site-title                   "My Awesome Blog"
 :author                       "Bob Bobbert"
 :description                  "This blog is awesome"
 :site-url                     "http://blogawesome.com/"
 :post-root                    "posts"
 :page-root                    "pages"
 :post-root-uri                "posts-output"
 :page-root-uri                "pages-output"
 :tag-root-uri                 "tags-output"
 :author-root-uri              "authors-output"
 :public-dest                  "public"
 :blog-prefix                  "/blog"
 :rss-name                     "feed.xml"
 :rss-filters                  ["cryogen"]
 :recent-posts                 3
 :post-date-format             "yyyy-MM-dd"
 :archive-group-format         "yyyy MMMM"
 :sass-src                     []
 :sass-path                    "sass"
 :theme                        "blue"
 :resources                    ["img"]
 :keep-files                   [".git"]
 :disqus?                      false
 :disqus-shortname             ""
 :ignored-files                [#"\.#.*" #".*\.swp$"]
 :previews?                    false
 :posts-per-page               5
 :blocks-per-preview           2
 :clean-urls                   :trailing-slash
 :collapse-subdirs?            false
 :hide-future-posts?           true
 :klipse                       {}
 :description-include-elements #{:p :h1 :h2 :h3 :h4 :h5 :h6}
 :debug?                       false}
```

For information about each key please see the ["Configuration"](http://cryogenweb.org/docs/configuration.html) portion of the Cryogen documentation site.

### Switching between Markdown and AsciiDoc

Cryogen comes with Markdown support as default. If you want to use AsciiDoc instead, open the `project.clj` in your created blog (e.g. `my-blog`), and change the line in `:dependencies` that says `cryogen-flexmark` to `cryogen-asciidoc` (and ensure the right version).
Instead of looking for files ending in `.md` in the `content/md` directory, the compiler will now look for files ending in `.asc` in the `content/asc` directory.

### Selecting a Theme

The Cryogen template comes with three themes in the `themes` folder. To change your blog's theme, change the value of the `:theme` key in `config.edn`. Note that the Nucleus theme is obtained from [downloadwebsitetemplates.co.uk](http://www.downloadwebsitetemplates.co.uk/template/nucleus/) that requires you to keep the footer, unless you make a donation on their website.

### Customizing Layouts

Cryogen uses [Selmer](https://github.com/yogthos/Selmer) templating engine for layouts. Please refer to its documentation to see the supported tags and filters for the layouts.

The layouts are contained in the `themes/{theme}/html` folder of the project. By default, the `base.html` layout is used to provide the general layout for the site. This is where you would add static resources such as CSS and JavaScript assets as well as define headers and footers for your site.

Each page layout should have a name that matches the `:layout` key in the page metadata and end with `.html`. Page layouts extend the base layout and should only contain the content relaveant to the page inside the `content` block.
For example, the `tag` layout is located in `tag.html` and looks as follows:

```xml
{% extends "/html/base.html" %}
{% block content %}
<div id="posts-by-tag">
    <h2>Posts tagged {{name}}</h2>
    <ul>
    {% for post in posts %}
        <li>
            <a href="{{post.uri}}">{{post.title}}</a>
        </li>
    {% endfor %}
    </ul>
</div>
{% endblock %}
```

### Code Syntax Highlighting

Cryogen uses [Highlight.js](https://highlightjs.org/) for code syntax highlighting. You can add more languages by replacing `themes/{theme}/js/highlight.pack.js` with a customized package from [here](https://highlightjs.org/download/).

The ` initHighlightingOnLoad` function is called in `themes/{theme}/html/base.html`.

```xml
<script>hljs.initHighlightingOnLoad();</script>
```

## Deploying Your Site

The generated static content will be found under the `public` folder. Simply copy the content to a static
folder for a server such as Nginx or Apache and your site is now ready for service.

A sample Nginx configuration that's placed in `/etc/nginx/sites-available/default` can be seen below:

```javascript
server {
  listen 80 default_server;
  listen [::]:80 default_server ipv6only=on;
  server_name localhost <yoursite.com> <www.yoursite.com>;

  access_log  /var/log/blog_access.log;
  error_log   /var/log/blog_error.log;

  location / {
    alias       /var/blog/;
    error_page  404 = /404.html;
  }
}
```

Simply set `yoursite.com` to the domain of your site in the above configuration and
ensure the static content is available at `/var/blog/`. Finally, place your custom error page
in the `/var/blog/404.html` file.

More information on deployment can be found [here](http://cryogenweb.org/docs/deploying-to-github-pages.html).

## Third Party Libraries

#### https://github.com/greywolve/cryogen-markdown-external

A Clojure library to provide Markdown rendering to the cryogen-core compiler by using an external command/program, such as pandoc.



## Some Sites Made With Cryogen

* [My personal blog](http://carmen.la/blog/archives)
* [Cryogen Documentation Site](http://cryogenweb.org)
* [Yogthos' blog](http://yogthos.net/)
* [Chad Stovern's blog](http://www.chadstovern.com)
* [szcz](http://www.szcz.org/)
* [ISvit blog](https://blog.isvit.info)
* [Lambda Funk](http://lambdafunk.com)
* [DDFT.wiki](https://ddft.wiki)
* [Ampersanda's blog](https://ampersanda.dev)
* [irrelevancy universe](https://irrelevancy.neocities.org)

## License

Copyright © 2014-2021 Carmen La

Distributed under the Eclipse Public License, the same as Clojure.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this
file. This change log follows the conventions of
[keepachangelog.com](http://keepachangelog.com/), with the addition of
sections for `Highlights`, `Breaking` and `Changes since x.y.z`.

## [Unreleased]

## [0.6.6] - 2021-10-14

### Highlights
- Support a `deps-new` template.

## [0.6.5] - 2021-10-07

### Highlights
- Bump up cryogen-core.

## [0.6.4] - 2021-10-04

### Highlights
- Update docs to the latest clj-new
### Breaking
- Fast compilation command line changed from `*-fast` to `*:fast`.

## [0.6.3] - 2020-12-07

### Highlights
- Allow fast compilation of changed files only. Opt-in. Requires
  `cryogen-core` version `0.4.0`.
- Support the use of `tools.deps` (aka clojure CLI) in addition to `lein`.

### Changes since 0.6.2
- [#226](https://github.com/cryogen-project/cryogen/pull/226) Prepare for release 0.6.3 (bombaywalla)
- [#225](https://github.com/cryogen-project/cryogen/pull/225) git-ignore lotus' sass-generated files (devurandom)
- [#221](https://github.com/cryogen-project/cryogen/pull/221) Support for deps tools instead of lein (holyjak)
- [#220](https://github.com/cryogen-project/cryogen/pull/220) Support for incremental compilation (holyjak)

## [0.6.2] - 2020-10-16

### Highlights
- Update to new version of the Markup protocol that allows for
  multiple file extensions for a particular markup.

### Breaking
- Requires latest versions of `cryogen-asciidoc` (at least 0.3.2),
  `cryogen-flexmark` (at least 0.1.4), and `cryogen-markdown` (at
  least 0.1.13).

### Changes since 0.6.1
- [#217](https://github.com/cryogen-project/cryogen/pull/217) Update README.md to current year (hewrin)
- [#216](https://github.com/cryogen-project/cryogen/pull/216) Fix typo in QuickStart template content (daemianmack)
- [#213](https://github.com/cryogen-project/cryogen/pull/213) Use Markup protocol defined in cryogen-core.markup (darth10)

## [0.6.1] - 2020-03-10

### Highlights
- Update versions of `cryogen-core` and `cryogen-flexmark`.

### Changes since 0.6.0
- [#206](https://github.com/cryogen-project/cryogen/pull/206) README: cryogen-markdown => cryogen-flexmark (daemianmack)

## [0.6.0] - 2020-01-20

### Highlights
- Switch to `cryogen-flexmark`.
- Handle case where `blog-prefix` is blank.

[Unreleased]: https://github.com/cryogen-project/cryogen/compare/0.6.6...HEAD

[0.6.4]: https://github.com/cryogen-project/cryogen/compare/0.6.5...0.6.6
[0.6.4]: https://github.com/cryogen-project/cryogen/compare/0.6.4...0.6.5
[0.6.4]: https://github.com/cryogen-project/cryogen/compare/0.6.3...0.6.4
[0.6.3]: https://github.com/cryogen-project/cryogen/compare/0.6.2...0.6.3
[0.6.2]: https://github.com/cryogen-project/cryogen/compare/0.6.1...0.6.2
[0.6.1]: https://github.com/cryogen-project/cryogen/compare/0.6.0...0.6.1
[0.6.0]: https://github.com/cryogen-project/cryogen/compare/0.1.0...0.6.0


src/leiningen/new/cryogen.clj
(ns leiningen.new.cryogen
  (:require [leiningen.new.templates :refer [renderer sanitize year ->files]]
            [leinjacker.utils :refer [lein-generation]]
            [leiningen.core.main :as main]
            [clojure.java.io :as io]))

(defn resource [r]
  (->> r (str "leiningen/new/cryogen/") (io/resource) (io/input-stream)))

(def render (renderer "cryogen"))

(defn check-lein-version []
  (if (< (lein-generation) 2)
    (throw (new Exception "Leiningen v2 is required..."))))

(def blue-theme
  [["themes/blue/html/archives.html" (render "themes/blue/html/archives.html")]
   ["themes/blue/html/author.html" (render "themes/blue/html/author.html")]
   ["themes/blue/html/base.html" (render "themes/blue/html/base.html")]
   ["themes/blue/html/home.html" (render "themes/blue/html/home.html")]
   ["themes/blue/html/page.html" (render "themes/blue/html/page.html")]
   ["themes/blue/html/post.html" (render "themes/blue/html/post.html")]
   ["themes/blue/html/post-content.html" (render "themes/blue/html/post-content.html")]
   ["themes/blue/html/previews.html" (render "themes/blue/html/previews.html")]
   ["themes/blue/html/tag.html" (render "themes/blue/html/tag.html")]
   ["themes/blue/html/tags.html" (render "themes/blue/html/tags.html")]
   ["themes/blue/html/404.html" (render "themes/blue/html/404.html")]
   ["themes/blue/css/screen.css" (render "themes/blue/css/screen.css")]
   ["themes/blue/js/highlight.pack.js" (render "themes/blue/js/highlight.pack.js")]])

(def blue-centered-theme
  [["themes/blue_centered/html/archives.html" (render "themes/blue_centered/html/archives.html")]
   ["themes/blue_centered/html/author.html" (render "themes/blue_centered/html/author.html")]
   ["themes/blue_centered/html/base.html" (render "themes/blue_centered/html/base.html")]
   ["themes/blue_centered/html/home.html" (render "themes/blue_centered/html/home.html")]
   ["themes/blue_centered/html/page.html" (render "themes/blue_centered/html/page.html")]
   ["themes/blue_centered/html/post.html" (render "themes/blue_centered/html/post.html")]
   ["themes/blue_centered/html/post-content.html" (render "themes/blue_centered/html/post-content.html")]
   ["themes/blue_centered/html/previews.html" (render "themes/blue_centered/html/previews.html")]
   ["themes/blue_centered/html/tag.html" (render "themes/blue_centered/html/tag.html")]
   ["themes/blue_centered/html/tags.html" (render "themes/blue_centered/html/tags.html")]
   ["themes/blue_centered/html/404.html" (render "themes/blue_centered/html/404.html")]
   ["themes/blue_centered/css/screen.css" (render "themes/blue_centered/css/screen.css")]
   ["themes/blue_centered/js/highlight.pack.js" (render "themes/blue_centered/js/highlight.pack.js")]])

(def nucleus-theme
  [["themes/nucleus/html/archives.html" (render "themes/nucleus/html/archives.html")]
   ["themes/nucleus/html/author.html" (render "themes/nucleus/html/author.html")]
   ["themes/nucleus/html/base.html" (render "themes/nucleus/html/base.html")]
   ["themes/nucleus/html/home.html" (render "themes/nucleus/html/home.html")]
   ["themes/nucleus/html/page.html" (render "themes/nucleus/html/page.html")]
   ["themes/nucleus/html/post.html" (render "themes/nucleus/html/post.html")]
   ["themes/nucleus/html/post-content.html" (render "themes/nucleus/html/post-content.html")]
   ["themes/nucleus/html/previews.html" (render "themes/nucleus/html/previews.html")]
   ["themes/nucleus/html/tag.html" (render "themes/nucleus/html/tag.html")]
   ["themes/nucleus/html/tags.html" (render "themes/nucleus/html/tags.html")]
   ["themes/nucleus/html/404.html" (render "themes/nucleus/html/404.html")]
   ["themes/nucleus/js/highlight.pack.js" (render "themes/nucleus/js/highlight.pack.js")]
   ["themes/nucleus/js/scripts.js" (render "themes/nucleus/js/scripts.js")]
   ["themes/nucleus/css/buttons.css" (render "themes/nucleus/css/buttons.css")]
   ["themes/nucleus/css/menu.css" (render "themes/nucleus/css/menu.css")]
   ["themes/nucleus/css/reset.css" (render "themes/nucleus/css/reset.css")]
   ["themes/nucleus/css/style.css" (render "themes/nucleus/css/style.css")]
   ["themes/nucleus/css/typography.css" (render "themes/nucleus/css/typography.css")]])

(def lotus-theme
  [["themes/lotus/config.edn" (render "themes/lotus/config.edn")]
   ["themes/lotus/README.md" (render "themes/lotus/README.md")]
   ["themes/lotus/css/_buttons.scss" (render "themes/lotus/css/_buttons.scss")]
   ["themes/lotus/css/_layout.scss" (render "themes/lotus/css/_layout.scss")]
   ["themes/lotus/css/_menu.scss" (render "themes/lotus/css/_menu.scss")]
   ["themes/lotus/css/_typography.scss" (render "themes/lotus/css/_typography.scss")]
   ["themes/lotus/css/_variables.scss" (render "themes/lotus/css/_variables.scss")]
   ["themes/lotus/css/blog.scss" (render "themes/lotus/css/blog.scss")]
   ["themes/lotus/css/lotus-highlightjs.min.css" (render "themes/lotus/css/lotus-highlightjs.min.css")]
   ["themes/lotus/css/normalize.css" (render "themes/lotus/css/normalize.css")]
   ["themes/lotus/html/404.html" (render "themes/lotus/html/404.html")]
   ["themes/lotus/html/archives.html" (render "themes/lotus/html/archives.html")]
   ["themes/lotus/html/author.html" (render "themes/lotus/html/author.html")]
   ["themes/lotus/html/base.html" (render "themes/lotus/html/base.html")]
   ["themes/lotus/html/home.html" (render "themes/lotus/html/home.html")]
   ["themes/lotus/html/page.html" (render "themes/lotus/html/page.html")]
   ["themes/lotus/html/post.html" (render "themes/lotus/html/post.html")]
   ["themes/lotus/html/post-content.html" (render "themes/lotus/html/post-content.html")]
   ["themes/lotus/html/prev-next.html" (render "themes/lotus/html/prev-next.html")]
   ["themes/lotus/html/previews.html" (render "themes/lotus/html/previews.html")]
   ["themes/lotus/html/tag.html" (render "themes/lotus/html/tag.html")]
   ["themes/lotus/html/tags.html" (render "themes/lotus/html/tags.html")]
   ["themes/lotus/img/black-lotus.svg" (render "themes/lotus/img/black-lotus.svg")]
   ["themes/lotus/img/icons.svg" (render "themes/lotus/img/icons.svg")]
   ["themes/lotus/img/white-lotus.svg" (render "themes/lotus/img/white-lotus.svg")]
   ["themes/lotus/js/highlight.pack.js" (render "themes/lotus/js/highlight.pack.js")]])

(def themes
  (concat blue-theme
          blue-centered-theme
          nucleus-theme
          lotus-theme))

(defn cryogen
  "Create a new Cryogen project"
  [name]
  (check-lein-version)
  (let [options {:name      name
                 :sanitized (sanitize name)
                 :year      (year)}]
    (main/info "Generating fresh 'lein new' Cryogen project.")
    (with-redefs [leiningen.new.templates/render-text (fn [text _] text)]
      (apply ->files
             options
             [".gitignore" (render "root/gitignore")]
             ["project.clj" (render "root/project.clj")]
             ["deps.edn" (render "root/deps.edn")]
             ;;static resources
             ["content/img/cryogen.png" (resource "content/img/cryogen.png")]
             ["content/css/example.css" (resource "content/css/example.css")]
             ["content/css/sassexample.scss" (resource "content/css/sassexample.scss")]
             ;;Markdown templates
             ["content/md/pages/about.md" (render "content/md/pages/about.md")]
             ["content/md/pages/another-page.md" (render "content/md/pages/another-page.md")]
             ["content/md/posts/2014-03-10-first-post.md" (render "content/md/posts/2014-03-10-first-post.md")]
             ["content/md/posts/2014-11-04-second-post.md" (render "content/md/posts/2014-11-04-second-post.md")]
             ["content/md/posts/2020-12-03-docs.md" (render "content/md/posts/2020-12-03-docs.md")]
             ;;Asciidoc templates
             ["content/asc/pages/adoc-page.asc" (render "content/asc/pages/adoc-page.asc")]
             ["content/asc/posts/2014-10-10-adoc-post.asc" (render "content/asc/posts/2014-10-10-adoc-post.asc")]
             ;;config
             ["content/config.edn" (render "content/config.edn")]
             ;;namespaces
             ["src/cryogen/core.clj" (render "src/cryogen/core.clj")]
             ["src/cryogen/server.clj" (render "src/cryogen/server.clj")]
             themes))))
